docs/design/roadmap.md

```diff
@@ -0,0 +1,168 @@
+
+## Provide access to "ambient parameters"
+
+*Planned for 0.9.0* ([#222](https://github.com/ksonnet/ksonnet/issues/222))
+
+Nearly all ksonnet prototypes expect a `namespace` parameter. However, in the Jsonnet code, there is currently no way of referencing the namespace of a given ksonnet environment.
+
+This leads to duplicate info—the namespace is saved both (1) in the environment metadata and (2) as an environment parameter.
+
+**There should be a way to surface these sorts of ambient parameters, something along the lines of `env.namespace`.**
+
+## Support JSON and YAML
+
+*Planned for 0.9.0* ([#240](https://github.com/ksonnet/ksonnet/issues/240))
+
+Right now only `*.jsonnet` files in the `components/` directory are recognized and used during commands like `ks apply` and `ks show`.
+
+This makes it more difficult for users with existing Kubernetes manifests to transition their app configurations into ksonnet. Although there is a workaround (converting YAML manifests to JSON and saving them as `*.jsonnet`, since JSON is a subset of Jsonnet), this is *not* an ideal workflow.
+
+**In future versions of ksonnet, users should be able to drop in a JSON, YAML, or Jsonnet file into `components/` and be able to deploy it with `ks apply`.** (Noting that only Jsonnet files can fully leverage parameters).
+
+## Reference components in other components
+
+*Planned for 1.0.0*
+
+Right now ksonnet allows users to parameterize literal values (e.g. setting the `image` of a `deployed-service` component). **However, there also needs to be a way for components to reference other components, ideally with a supporting CLI command**.
+
+For example, it might be useful for the Deployment of a `redis-stateless` component to know about a shared `configMap` component. The actual implementation of this is up in the air--whether it looks like component references via `ks param` or something like `ks add <component1> <component2>`.
+
+Either way, it is important for this command to facilitate *incremental* additions in an intuitive way. After all, when developing Kubernetes apps, it's rare to know all of the various components you need from the beginning.
+
+## Establish clear `ks apply` semantics
+
+*Planned for 0.9.0* ([#200](https://github.com/ksonnet/ksonnet/issues/200))
+
+Currently, `ks apply` garbage collects *all* previously created API resources from your ksonnet app, before recreating everything defined in `components/`. This has a lot of the advantages of declarative configuration:
+* Deleting a component in your ksonnet app is actually reflected server-side
+* It's generally easier for users to reason through what will run on their cluster
+
+However, it also leads to bugs where `ks apply` clobbers fields that ought to have been retained. This means that users get a different `nodePort` each time they `ks apply` a `ClusterIP` Service, and that changes from horizontal pod autoscaling are lost.
```

Comment:
Just curious but is `ks apply` with the AnnotationGcStrategy meant for "recreating" every object managed by ksonnet, hence `nodePort` changes each time you run `ks apply` as noted here, according to code [here](https://github.com/ksonnet/ksonnet/blob/10bbf40f4383b27f72c5fa1aeff24e339729f010/pkg/kubecfg/apply.go#L88) and [there](https://github.com/ksonnet/ksonnet/blob/10bbf40f4383b27f72c5fa1aeff24e339729f010/cmd/apply.go#L119-L122)?

Reply:
The goal here is to maintain items in a declarative fashion. If they don't exist, create them.  If they do exist and aren't needed any more, delete them.